### PR TITLE
[WIP]Pagination for Assets API implemented

### PR DESCRIPTION
**Pull request by deepusundar**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### CODE REVIEW

File: core/jazz_assets/components/crud/getList.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
+ /**
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: header?

File: core/jazz_assets/components/crud/getList.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+ module.exports = (query, asset_table, onComplete) => {
+     // initialize dynamodb
+     var docClient = utils.initDocClient();
+
+     var filter = "";
+     var insertAndString = " AND ";
+     var attributeValues = {};
+     var attributeNames = {};
+
+     var params = {
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: try using const/lets. ES6 stuff. Please update in other places as well

File: core/jazz_assets/components/crud/getList.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+         ExpressionAttributeNames: {
+             "#d": "DOMAIN"
+         }
+     };
+
+     var keys_list = global.global_config.ASSET_SEARCH_OPTIONAL_FILTER_PARAMS;
+
+     // Generate filter string
+     keys_list.forEach((key) => {
+            if(key != "limit" && key != "offset") { // LIMIT is a reserved keyword
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: indentation - if (
also, !==

File: core/jazz_assets/components/crud/getList.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        pagination.limit = global.global_config.PAGINATION_DEFAULTS.max_limit
+     }
+     pagination.offset = query.offset ||  global.global_config.PAGINATION_DEFAULTS.offset;
+
+     logger.debug("Query params generated from the seach request: " + JSON.stringify(params));
+
+     var items = [];
+     var queryExecute = (onComplete) => {
+         docClient.query(params, (err, data) => {
+             var count
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: no semi-colon.

File: core/jazz_assets/components/crud/getList.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      @module: getList.js
+      @description: CRUD functions for asset catalog
+      @author:
+      @version: 1.0
+  **/
+
+ const utils = require("../utils.js"); //Import the utils module.
+ const logger = require("../logger.js")(); //Import the logging module.
+
+ module.exports = (query, asset_table, onComplete) => {
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: There is too much going on in this handler. Can you modularize the code (create some functions)? and make it more readable and testable?

File: core/jazz_assets/components/crud/getList.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+     var keys_list = global.global_config.ASSET_SEARCH_OPTIONAL_FILTER_PARAMS;
+
+     // Generate filter string
+     keys_list.forEach((key) => {
+            if(key != "limit" && key != "offset") { // LIMIT is a reserved keyword
+                var key_name = utils.getDatabaseKeyName(key);
+
+                if (query[key] && key_name) {
+                    filter = filter + key_name + " = :" + key_name + insertAndString;
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: consider using template literals(ES6 feature)

File: core/jazz_assets/components/crud/getList.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+                    params.ExpressionAttributeValues[":" + key_name] = query[key];
+                }
+            }
+     });
+
+     filter = filter.substring(0, filter.length - insertAndString.length); // remove the " AND " at the end
+
+     if (filter) {
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: you are checking for if(filter here) what if filter is undefined in line 48 itself?

File: core/jazz_assets/components/crud/getList.js [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+         params.FilterExpression = filter;
+     }
+
+     let pagination = {}
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: please add semi colon. also check other places in the code

File: core/jazz_assets/components/crud/getList.js [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+             var count
+             if (err) {
+                 onComplete(err, null);
+             } else {
+                 items = items.concat(data.Items);
+                 if (data.LastEvaluatedKey) {
+                     params.ExclusiveStartKey = data.LastEvaluatedKey;
+                     queryExecute(onComplete);
+                 } else {
+
```
Comment (svidhani) [inline core/jazz_assets/components/crud/getList.js]: remove extra line

File: core/jazz_assets/components/crud/getList.js (hunk 1 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: added

File: core/jazz_assets/components/crud/getList.js (hunk 8 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: updated

File: core/jazz_assets/components/crud/getList.js (hunk 9 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: done

File: core/jazz_assets/components/crud/getList.js (hunk 2 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: updated

File: core/jazz_assets/components/crud/getList.js (hunk 3 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: updated

File: core/jazz_assets/components/crud/getList.js (hunk 4 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: updated

File: core/jazz_assets/components/crud/getList.js (hunk 5 above)
Comment (deepusundar) [reply core/jazz_assets/components/crud/getList.js]: Now this is consistent across all services. The only logic it does is, create/format a get payload and hit the DB

File: core/jazz_metrics/index.js [hunk 10]
```diff
@@ -28,73 +28,73 @@ const request = require('request');
 const utils = require("./components/utils.js"); //Import the utils module.
 const validateUtils = require("./components/validation.js");
 
-function handler (event, context, cb) {
-  var errorHandler = errorHandlerModule();
-  var config = configObj.getConfig(event, context);
+function handler(event, context, cb) {
```
Comment (suryajak) [inline core/jazz_metrics/index.js]: Looks like you are overriding the editorconfig, is this intentional?

Comment (deepusundar) [reply core/jazz_metrics/index.js]: Not really. But I applied a formatting through VSCode formatter. Would that change anything ?

File: core/jazz_assets/index.js [hunk 11]
```diff
@@ -80,6 +80,18 @@ function handler(event, context, cb) {
               logger.error("update error:" + JSON.stringify(error));
               handleResponse(error, null, event.path);
             });
+        } else if (event.method === 'GET' && !assets_id) { // Call GetList //TODO
+          logger.info('GET assets list by service name and domain');
```
[the remark this answers is not in the document]
Comment (deepusundar) [reply core/jazz_assets/index.js]: Yes, we can make it debug too. Will update this for this file.

File: core/jazz_assets/index.js [hunk 12]
```diff
@@ -80,6 +80,18 @@ function handler(event, context, cb) {
...
+          var query = event.query;
+          exportable.processGetList(query, asset_table)
+            .then(res => {
+              logger.info("GET assets list by service name and domain:" + JSON.stringify(res));
```
[the remark this answers is not in the document]
Comment (deepusundar) [reply core/jazz_assets/index.js]: done

File: core/jazz_assets/index.js [hunk 13]
```diff
@@ -112,20 +124,7 @@ function handler(event, context, cb) {
...
-              logger.info("search asset result:" + JSON.stringify(res));
-              handleResponse(null, res, assets_data);
-            })
-            .catch(error => {
-              logger.error("create asset error:" + JSON.stringify(error));
-              handleResponse(error, null, assets_data);
-            });
-        } else {
+        else {
           logger.error(JSON.stringify(event));
```
[the remark this answers is not in the document]
Comment (deepusundar) [reply core/jazz_assets/index.js]: updated

File: core/jazz_assets/components/utils.js [hunk 14]
```diff
@@ -195,11 +195,24 @@ var toLowercase = (input_data) => {
 	return asset_data;
 };
 
+var paginateUtil = (data, limit, offset) => {
+    var newArr = [];
+    if (offset > data.length || offset == data.length || limit === 0) {
```
[the remark this answers is not in the document]
Comment (deepusundar) [reply core/jazz_assets/components/utils.js]: I have kept the utility module as it is.  If there is a bug or improvement in modules, lets address it across board.

File: core/jazz_metrics/index.js (hunk 10 above)
Comment (suryajak) [reply core/jazz_metrics/index.js]: Are you using a VSCode extension? [This](https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig) extension works with editorconfig. Once this extension is in, open your target file and use the shortcut `Shift + Option + F` for the repo formats to be applied.

Comment (deepusundar) [reply core/jazz_metrics/index.js]: Applied the repo config with editorConfig plugin

**Pull request merged**